cpp/src/arrow/dataset/dataset_writer.cc

```diff
@@ -0,0 +1,569 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/dataset/dataset_writer.h"
+
+#include <list>
+#include <mutex>
+#include <unordered_map>
+
+#include "arrow/filesystem/path_util.h"
+#include "arrow/record_batch.h"
+#include "arrow/result.h"
+#include "arrow/util/future.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/map.h"
+#include "arrow/util/string.h"
+
+namespace arrow {
+namespace dataset {
+
+namespace {
+
+constexpr util::string_view kIntegerToken = "{i}";
+
+class DatasetWriterStatistics {
+ public:
+  DatasetWriterStatistics(uint64_t max_rows_in_flight, uint32_t max_files_in_flight)
+      : max_rows_in_flight_(max_rows_in_flight),
+        max_files_in_flight_(max_files_in_flight) {}
```

Comment:
It is bizarre IMHO to have a single class managing different independent limits (you may want to write a new batch to an already open file even if the file limit was reached).

Why not separate classes? Also, perhaps give them more meaningful names e.g. `RowsLimiter` and `OpenFilesLimiter`.

Reply:
I split this logic into a generic Throttle class and now have `open_files_throttle_` and `rows_in_flight_throttle_`